Add CellAssign process

sjspielman opened the pull request:
Towards #415 

The next step is running cell assign, done here! I think given all the nice organizational changes in #476 & #477 this is pretty straightforward for review. The one additional thing to note is that I realized we probably need to be including `singler_reference_name` in the same way that we include `cellassign_reference_name`. This wasn't previously needed since we had been doing singler celltyping _and adding to SCE object_ in the same process, where the reference name was accessible from the trained singler model itself. Now that we will be adding singler & cellassign to the final step all at once in the last process, we'll need to grab both reference names for input.

I also have one discussion-y comment to bring up at this point - my memory from our figjam session 🎸  was that we had decided not to keep using indicator variables in meta, e.g. `meta.has_singler=true` (edit lowercase true, this isn't R). I am not seeing any notes that explicitly say this though, so let's make sure I am remembering correctly! We can/should certainly still add relevant information to _SCE metadata_ as we had been doing, but with our current celltyping workflow structure I don't really see the need for anything in the json metadata unless I am missing something?

Noting the next PR will write the process & R script for adding celltypes into the processed SCE.

jashapiro reviewed (commented):
Looks good. I suggested a few changes: 

- Using static filenames within the "bash script" of the process. Since these files are isolated and never "escape" the process, it isn't worth assigning them library-specific names.
- Similarly, for checkpoint files, it doesn't seem worth the effort to give them special names: the directory they are in will already encode specificity to the library
- Removing the `singler_reference_name` variable. We didn't have it before because it is already in the `.rds` files. 
  - Speaking of the `.rds` files: If we keep this structure where we are not passing the singler results directly along to cellassign, we might want to trim the output `singler_results.rds ` file in `classify_SingleR.R` output to save space. This would be a separate PR, and I would probably want to make trimming an option at the script level (which we would invoke in the process).
- Finally: no need to add extra values to the input channel for `assign_celltypes`: As long as they are in `meta`, we can access that from within the process.

jashapiro commented:
> I also have one discussion-y comment to bring up at this point - my memory from our figjam session 🎸 was that we had decided not to keep using indicator variables in meta, e.g. `meta.has_singler=true` (edit lowercase true, this isn't R). I am not seeing any notes that explicitly say this though, so let's make sure I am remembering correctly! We can/should certainly still add relevant information to _SCE metadata_ as we had been doing, but with our current celltyping workflow structure I don't really see the need for anything in the json metadata unless I am missing something?

The presence of a reference file(s) for singleR and/or cellassign will be enough to say whether that processing has happened. There should not be any need for a separate boolean value.

jashapiro commented:
> * Removing the `singler_reference_name` variable. We didn't have it before because it is already in the `.rds` files.

I realized that I am not 100% sure whether the reference name _is_ in the output `.rds` object. It was there when we included the full SCE object, but maybe not now? I'd still say it is better if we can include it in the object itself, rather than creating an extra `meta` variable and associated column that we would have to add to the input table.

sjspielman commented:
A couple thoughts on reference names - 

1. The singler trained model name _and_ the cellassign reference names are both present in their respective file names, .e.g. `BlueprintEncodeData_model.rds` & `PanglaoDB-blood.tsv`. So, since these filenames are in `meta`, I guess we don't need to pass either the singler or cellassign reference names in separately. We could extract this info from filenames with some string manipulation. (Not loving that we have `_` vs `-`, though..). 
3. In terms of adding the singler model name to the singler object saved in`rds`, I'm not sure how that would play out - my idea was to keep that as the exact structure that singler exports so that people can actually use it if they want to with singler's native functions, e.g. heatmap. It's a `DataFrame`, so we'd have to add a column with model name which should be fine unless there is crazy strict checking in singler's functions. But, if we do my point in 1^ , then this isn't needed.

sjspielman commented:
> I realized that I am not 100% sure whether the reference name is in the output .rds object. It was there when we included the full SCE object, but maybe not now? 

It is not in the output rds from the singler projcess at this point:
https://github.com/AlexsLemonade/scpca-nf/blob/fb5f3bf1d9dad7533046cf05230a5cf0b3ddabaf/modules/classify-celltypes.nf#L25

sjspielman commented on modules/classify-celltypes.nf:
```diff
@@ -100,31 +121,34 @@ workflow annotate_celltypes {
          it.scpca_project_id,
          // singler model file
          Utils.parseNA(it.singler_ref_file) ? "${params.singler_models_dir}/${it.singler_ref_file}" : null,
+         // singler reference name 
+         Utils.parseNA(it.singler_ref_name),       
```
just noting i'll circle back to this suggestion after final.final discussion in the PR!

jashapiro replied:
> The reason we didn't have this before is because it is encoded in the rds file of the SingleR model. Since the cellassign input is just a tsv, we don't have that info.

I was a bit confused: the reference name was being parsed out of the file name: https://github.com/AlexsLemonade/scpca-nf/blob/67e175bdbbfb314e0870a1b832d48001bfa87d9e/bin/classify_SingleR.R#L78

Looking back at https://github.com/AlexsLemonade/scpca-nf/pull/402, it seems like we added that variable to the table mostly for convenience, but we could have parsed it out of the file name. It seems like we are using a shorthand at the moment, but the full file name (less `.tsv`) should probably be included as the name (i.e. "PanglaoDB-blood" rather than "blood"). In nextflow this can be done with `file("path/my_model.tsv").baseName` which strips both the directory and the extension.

And fun addition, if we want to do the same string parsing for the singler model files, which end with "_model.rds", we could do: `file("path/my_model.tsv").baseName - '_model'` which is so odd, but it works. If you want to be safer, use a little regex instead: `file("path/my_model.tsv").baseName - ~'_model$'`. Yes, you can make a regex by just adding a ~ before a quote. Groovy is so strange, but kind of cool sometimes.

jashapiro replied:
> just noting i'll circle back to this suggestion after final.final discussion in the PR!

My current suggestion is to punt this to a separate PR: it isn't really part of getting the cellassign process done.

sjspielman commented:
> But we can add it quite trivially, as DataFrame objects have metadata slots just like the SCE. So we could add in something like metadata(singler_result)$singler_reference <- singler_reference_name to store the reference name along with the results itself. 

Right, I would just want to make sure this doesn't cause singler's native functions to break, which it really shouldn't. Let me double check though...

jashapiro commented:
> It is possible for the `meta` object to become disconnected from the file name/reference that was actually used if we are not careful, but I expect that the `repeat_celltyping` logic should handle that when we add it.

One way to handle this would be to (optionally) pass in the `spca-meta.json` files that are in the cellassign and singler directories

sjspielman commented:
> As long as we are adding in metadata, I'm fine with passing in the reference names for both cellassign and singler directly as an argument to the assign script. The only thing we have to be cautious about (and we have to do this anyway) is that we need to make sure we are passing in the correct value.

Feels like this could serve as an extra check as well - we'd want to make sure the singler reference name that gets passed in is the same as the name recorded in the singler DataFrame's metadata.

sjspielman commented:
Changes since last time:

- https://github.com/AlexsLemonade/scpca-nf/pull/479/commits/9de11fc74436ccf2ae5229bf38f55fd158bdf39b: I got rid of all tracking of the reference names in favor of "regexing them," if you will..., out of the reference/model file names directly.
- https://github.com/AlexsLemonade/scpca-nf/pull/479/commits/09565d13607fefecf064070d7db20b97e0ecaf3a: I updated the SinglerR script to add the reference name as DataFrame metadata, with a couple checks to go with it

sjspielman commented on bin/classify_SingleR.R:
```diff
@@ -55,19 +56,28 @@ if (!file.exists(opt$sce_file)) {
+}
+
+# get & check reference name
+reference_name <- stringr::str_remove(singler_model_file, "_model.rds")
+if (reference_name == "") {
+  stop(glue::glue("Provided model file name must be formatted as `<model_name>_model.rds`"))
+}
```
Hm, maybe this isn't the right stringr function. It's fine for pulling out the model name, but not for _checking_ the model name. Maybe we want `string::str_match()`?

jashapiro commented on bin/classify_SingleR.R:
```diff
@@ -55,19 +56,28 @@ if (!file.exists(opt$sce_file)) {
   stop(glue::glue("Provided model file {singler_model_file} is missing."))
 }
+if (!(stringr::str_ends(singler_model_file, "_model.rds"))) {
+  stop(glue::glue("Provided model file {singler_model_file} must end in .rds."))
```
```suggestion
  stop(glue::glue("Provided model file {singler_model_file} must end in '_model.rds.'"))
```

jashapiro commented on bin/classify_SingleR.R:
```diff
@@ -55,19 +56,28 @@ if (!file.exists(opt$sce_file)) {
+}
+
+# get & check reference name
+reference_name <- stringr::str_match(singler_model_file, "(\\w+)_model.rds")[2]
+if (reference_name == "") {
+  stop(glue::glue("Provided model file name must be formatted as `<model_name>_model.rds`"))
+}
```
I think the `str_remove` is more clear and simpler.

Your already checked above that the file had the right ending, so the second check here will only be triggered if the full file name was "_model.rds". 
```suggestion
reference_name <- stringr::str_remove(singler_model_file, "_model.rds$")
if (reference_name == "") {
  stop(glue::glue("Provided model file name must be formatted as `<model_name>_model.rds`"))
}
```

jashapiro commented on bin/classify_SingleR.R:
```diff
@@ -102,6 +116,7 @@ readr::write_tsv(
 )
 
 # next, the full result to a compressed rds
+
```
```suggestion
```

jashapiro commented on modules/classify-celltypes.nf:
```diff
@@ -101,30 +120,27 @@ workflow annotate_celltypes {
           meta.singler_dir = "${meta.celltype_publish_dir}/${meta.library_id}_singler";
           meta.cellassign_dir = "${meta.celltype_publish_dir}/${meta.library_id}_cellassign";
-          meta.singler_model_file = singler_model_file;
+          meta.singler_model_file = singler_model_file; 
```
```suggestion
          meta.singler_model_file = singler_model_file;
```

jashapiro commented on modules/classify-celltypes.nf:
```diff
@@ -133,23 +149,50 @@ workflow annotate_celltypes {
       // perform singleR celltyping and export results
       classify_singler(singler_input_ch.do_singler)
-
+      
```
```suggestion

```

jashapiro commented on modules/classify-celltypes.nf:
```diff
@@ -101,30 +120,27 @@ workflow annotate_celltypes {
-          meta.singler_model_file = singler_model_file;
+          meta.singler_model_file = singler_model_file; 
           meta.cellassign_reference_file = cellassign_reference_file;
-          meta.cellassign_reference_name = cellassign_reference_name;
```
I assume that by removing this here, you plan to add some code to `add_celltypes_to_sce` to pull out the cellassign reference name?

sjspielman replied:
Yes, that was the idea!

jashapiro reviewed (approved):
This looks fine, with a few small suggestions (some just whitespace). 
Any concerns related to tracking reference file names can be settled when we are actually using that part.

sjspielman commented:
Just for posterity...

- stub failed for https://github.com/AlexsLemonade/scpca-nf/pull/479/commits/3c2715cbb808124be2b41b4b6b5a38ec393b6aed, but none of these changes really should have caused that?
- Locally, stub ran fine for me at that commit!
- I retriggered the CI stub check and now it's passing 🤷‍♀️
